### ISSUE TITLE
Erroneous mixin_inference_no_matching_class error in Flutter

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#32452]
I encountered an mixin_inference_no_matching_class error in Flutter:

> ERROR: flutter/lib/src/rendering/proxy_box.dart:37
> Type parameters could not be inferred for the mixin 'RenderProxyBoxMixin' because the base class does not implement the mixin's supertype constraint 'RenderObjectWithChildMixin<T>' #mixin_inference_no_matching_class

Here are those two classes, [RenderProxyBox](https://github.com/flutter/flutter/blob/fbf68bbc6daa7a4f0b6f7b738b0134a318c4974b/packages/flutter/lib/src/rendering/proxy_box.dart#L37) and [RenderProxyBoxMixin](https://github.com/flutter/flutter/blob/fbf68bbc6daa7a4f0b6f7b738b0134a318c4974b/packages/flutter/lib/src/rendering/proxy_box.dart#L56):

```dart
class RenderProxyBox extends RenderBox
    with RenderObjectWithChildMixin<RenderBox>, RenderProxyBoxMixin<RenderBox> { ... }

@optionalTypeArgs
abstract class RenderProxyBoxMixin<T extends RenderBox> extends RenderBox
    with RenderObjectWithChildMixin<T> { ... }
```

So if a RenderProxyBoxMixin<T>'s T is RenderBox, then RenderProxyBox must mixin `RenderObjectWithChildMixin<RenderBox>`, which is exactly what it does... bug?

**Comment by leafpetersen** [comment]
So is this an analyzer bug?

**Comment by srawlins** [comment]
I think @stereotype441 and I decided this is not a bug. Is it OK to close this, Paul?

**Comment by leafpetersen** [comment]
I'm a little confused.  The code in the first comment seems to match the code in siva's workaround.  Was the error from the code before or after?  The code after doesn't involve any inference does it?

**Comment by leafpetersen** [comment]
It's not clear to me why siva's fix was needed, but maybe I'm missing something - reinterpreting the class syntax into sane mixin syntax gives me a headache.

**Comment by srawlins** [comment]
Sorry, yeah. I hit this bug when synced to a flutter commit _before_ the code I found on GitHub at master (i.e. _before_ Siva's fix). So what I copy/pasted was _not_ the code that triggered the bug.

I'm not clear either, but I think Paul _also_ expressed unhappiness that Siva's fix was needed. I think he called it a "work around." So I'll let Paul comment on whether this is a bug or not. I think the code when I hit the bug was:

```dart
class RenderProxyBox extends RenderBox
    with RenderObjectWithChildMixin<RenderBox>, RenderProxyBoxMixin { ... }

@optionalTypeArgs
abstract class RenderProxyBoxMixin<T extends RenderBox> extends RenderBox
    with RenderObjectWithChildMixin<T> { ... }
```

which Siva fixed to:

```dart
class RenderProxyBox extends RenderBox
    with RenderObjectWithChildMixin<RenderBox>, RenderProxyBoxMixin<RenderBox> { ... }

@optionalTypeArgs
abstract class RenderProxyBoxMixin<T extends RenderBox> extends RenderBox
    with RenderObjectWithChildMixin<T> { ... }
```

**Comment by leafpetersen** [comment]
Ok, so maybe this is just a bug with inferring from multiple superclass constraints?

**Comment by stereotype441** [comment]
Yes, I believe this is a bug with inferring from multiple superclass constraints.  It's complicated because there are many versions of the analyzer around with different degrees of fixedness, and there are two code paths through the analyzer with different degrees of fixedness.  And then of course there's Siva's workaround.

Let's start with the two code paths: they are the (new) analysis driver and the (old) task/context model.  AFAIK, everything uses the new analysis driver except DDC and a few use cases of flutter analyze (see https://github.com/dart-lang/sdk/issues/32421#issuecomment-370597792).  So if you're running the analyzer on the command line, via the analysis server, or through the Google internal build process, you're getting the new analysis driver.  But if you're running flutter analyze you might be getting the old task/context model.

Now let's talk about how that interplays with analyzer versions.  The new analysis driver handles this case properly on bleeding edge master (I just tested it).  I believe it has worked properly since 11cbdbe26e2ebc0367f3422a7ad71bb9ab03dd96, though I haven't tested that directly.  11cbdbe26e2ebc0367f3422a7ad71bb9ab03dd96 was released in 2.0.0-dev.33.0.  So if you're running an older version than that, you're expected to see problems.  Note that the internal codebase is running earlier code than that, so if you're using the internal version of the analyzer you're expected to see problems.

The old task/context model does not handle this case properly because it doesn't do mixin inference at all.  In b5b8080b2cfb1e4943a19787ee96548a4a604f7b I disabled some of the old task/context model's error checking to try to prevent the most obvious errors this causes.  In https://dart-review.googlesource.com/c/sdk/+/45524 I tried to disable even more of the error checking.  If you are seeing this error and you are using the old task/context model, https://dart-review.googlesource.com/c/sdk/+/45524 ought to make it go away.  But it's possible that there will be other follow-on errors that https://dart-review.googlesource.com/c/sdk/+/45524 isn't able to suppress.

Finally, Siva's workaround is expected to make the error go away in all circumstances, because it supplies the mixin type parameter directly, avoiding the need for inference.

I hope that helps somewhat.  Sam, it wasn't clear to me whether you saw the problem while trying to roll the SDK internally, while trying to roll analyzer code to Flutter, or under some other circumstance.  Can you clarify?  I'm happy to look in more detail if you send me repro instructions, but this is my last day in Aarhus and I'm leaving at 3pm (6am Pacific time), so unless you happen to still be awake it's unlikely that I'll be able to help today.  Next week I'm in London, so if necessary I can dig into this Monday before the workday starts for anyone in the US.

**Comment by srawlins** [comment]
> So if you're running the analyzer on the command line, via the analysis server, or through the Google internal build process

Yes, internal build process

>  11cbdbe was released in 2.0.0-dev.33.0. So if you're running an older version than that, you're expected to see problems.

Yes, this was testing -dev.32.0.

So I think in my case, it sounds like this has been fixed (in -dev.33.0). I did not encounter the bug in the old task/context model, so I don't think it needs to remain open just for that case. Probably close then as fixed in -dev.33.0?

**Comment by leafpetersen** [comment]
Ok, that clarifies somewhat.  I'm not entirely sure why Sam was seeing this, but Siva's fix would have been because flutter analyze is still using the task model.

https://github.com/dart-lang/sdk/issues/32478

I'll go ahead and close this now unless there's anything else to track.
